File: 10_laba/src/by/bntu/fitr/java/inGoslingsFootsteps/Lab10/model/logic/FlightController.java
package by.bntu.fitr.java.inGoslingsFootsteps.Lab10.model.logic;

import by.bntu.fitr.java.inGoslingsFootsteps.Lab10.model.entity.AirTicketOfficeContext;
import by.bntu.fitr.java.inGoslingsFootsteps.Lab10.model.entity.Customer;
import by.bntu.fitr.java.inGoslingsFootsteps.Lab10.model.entity.Flight;
import by.bntu.fitr.java.inGoslingsFootsteps.Lab10.model.exception.BookingException;
import org.apache.log4j.Logger;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.Arrays;

public class FlightController {
    private static final Logger LOG;

    private static AirTicketOfficeContext context;

    static{
        LOG = Logger.getLogger(by.bntu.fitr.java.inGoslingsFootsteps.Lab10.model.logic.FlightController.class);
    }

    public FlightController() throws FileNotFoundException {
        context = AirTicketOfficeContext.getContext();
    }

    public AirTicketOfficeContext getAvailableFlight(Customer customer) throws BookingException {
        AirTicketOfficeContext availableFlight = new AirTicketOfficeContext();

        for (int i = 0; i < context.sizeOfFlight(); i++) {
            boolean isEmpty = true;
            String[] customersId = context.getFlightByIndex(i).getCustomersId();
            if(context.getFlightByIndex(i).getCustomersId().length == 0) {
                availableFlight.setFlights(addFlightToArray(availableFlight, i).getFlights());
                isEmpty = false;
            }
            for(int j =0; j < context.getFlightByIndex(i).getCustomersId().length; j++) {
                if (customer.getId().equals(customersId[j])) {
                    isEmpty = false;
                }
            }
            if(isEmpty){
                availableFlight.setFlights(addFlightToArray(availableFlight, i).getFlights());
            }
        }
        if (availableFlight.getFlights().length == 0) {
            throw new BookingException("You haven't available flight. Sorry(");
        }
        LOG.info("Got list of available flights");
        return availableFlight;
    }

    public AirTicketOfficeContext getFlightsBookedByCustomer(Customer customer) throws BookingException {
        AirTicketOfficeContext occupiedCars = new AirTicketOfficeContext();
        for (int i = 0; i < context.sizeOfFlight(); i++) {
            String[] customersId = context.getFlightByIndex(i).getCustomersId();
            for (int j = 0; j < context.getFlightByIndex(i).getCustomersId().length; j++) {
                if (customer.getId().equals(customersId[j])) {
                    occupiedCars.setFlights(addFlightToArray(occupiedCars, i).getFlights());
                    break;
                }
            }
        }
        if (occupiedCars.getFlights().length == 0) {
            throw new BookingException("You haven't booked flight.");
        }
        LOG.info("Got list of booked by customer flights");
        return occupiedCars;
    }
    private AirTicketOfficeContext addFlightToArray(AirTicketOfficeContext airTicketOfficeContext, int index) {
        airTicketOfficeContext.setFlights(Arrays.copyOf(airTicketOfficeContext.getFlights(),
                airTicketOfficeContext.getFlights().length + 1));
        airTicketOfficeContext.getFlights()[airTicketOfficeContext.getFlights().length - 1] =
                context.getFlightByIndex(index);
        return airTicketOfficeContext;
    }

    public void bookTicket(Flight flight, Customer customer) throws IOException {
        flight.addCustomersId(customer);
        context.saveChanges();
        LOG.info("Data saved");
    }

    public void returnTicket(Flight flight, Customer customer) throws IOException {
            String[] customersId = flight.getCustomersId();
            for(int i =0; i < customersId.length; i++) {
                if (customer.getId().equals(customersId[i])) {
                    flight.removeCustomerId(i);
                }
            }
        context.saveChanges();
        LOG.info("Data saved");
    }
}

File: 10_laba/C##/AirTicketOffice/AirTicketOffice/by/bntu/fitr/java/ingoslingsfootsteps/model/logic/FlightController.cs
using AirTicketOffice.by.bntu.fitr.java.ingoslingsfootsteps.model.entity;
using AirTicketOffice.by.bntu.fitr.java.ingoslingsfootsteps.model.exception;
using NLog;
using NLog.Fluent;
using System;
using System.Collections.Generic;
using System.Text;

namespace AirTicketOffice.by.bntu.fitr.java.ingoslingsfootsteps.model.logic
{
    public class FlightController
    {
        private static readonly Logger logger = LogManager.GetCurrentClassLogger();

        private static AirTicketOfficeContext context;

        public FlightController()
        {
            context = AirTicketOfficeContext.Context;
        }

        public AirTicketOfficeContext GetAvailableFlight(Customer customer)
        {
            AirTicketOfficeContext availableFlight = new AirTicketOfficeContext();

            for (int i = 0; i < context.Flights.Length; i++)
            {
                bool isEmpty = true;
                string[] customersId = context.GetFlightByIndex(i).CustomersId;
                 if (context.GetFlightByIndex(i).CustomersId.Length == 0)
                {
                    availableFlight.Flights = AddFlightToArray(availableFlight, i).Flights;
                    isEmpty = false;
                }
                for (int j = 0; j < context.GetFlightByIndex(i).CustomersId.Length; j++)
                {
                    if (customer.Id.Equals(customersId[j]))
                    {
                        isEmpty = false;
                    }
                }
                if (isEmpty)
                {
                    availableFlight.Flights = AddFlightToArray(availableFlight, i).Flights;
                }
            }
            if (availableFlight.Flights.Length == 0)
            {
                throw new BookingException("You haven't available flight. Sorry(");
            }
            logger.Info("Got list of available flights");
            return availableFlight;
        }

        public AirTicketOfficeContext GetFlightsBookedByCustomer(Customer customer)
        {
            AirTicketOfficeContext occupiedCars = new AirTicketOfficeContext();
            for (int i = 0; i < context.SizeOfFlight(); i++)
            {
                String[] customersId = context.GetFlightByIndex(i).CustomersId;
                if (context.GetFlightByIndex(i).CustomersId.Length == 0)
                    continue;
                for (int j = 0; j < context.GetFlightByIndex(i).CustomersId.Length; j++)
                {
                    if (customer.Id.Equals(customersId[j]))
                    {
                        occupiedCars.Flights = AddFlightToArray(occupiedCars, i).Flights;
                        break;
                    }
                }
            }
            if (occupiedCars.Flights.Length == 0)
            {
                throw new BookingException("You haven't booked flight.");
            }
            logger.Info("Got list of booked by customer flights");
            return occupiedCars;
        }
        private AirTicketOfficeContext AddFlightToArray(AirTicketOfficeContext airTicketOfficeContext, int index)
        {
            Flight[] flights = airTicketOfficeContext.Flights;
            Array.Resize(ref flights, flights.Length + 1);
            flights[flights.Length - 1] = context.GetFlightByIndex(index);
            airTicketOfficeContext.Flights = flights;
            return airTicketOfficeContext;
        }

        public void BookTicket(Flight flight, Customer customer)
        {
            flight.AddCustomersId(customer);
            logger.Info("Data saved");
            context.SaveChanges();
        }

        public void ReturnTicket(Flight flight, Customer customer)
        {
            string[] customersId = flight.CustomersId;
            for (int i = 0; i < customersId.Length; i++)
            {
                if (customer.Id.Equals(customersId[i]))
                {
                    flight.RemoveCustomerId(i);
                }
            }
            logger.Info("Data saved");
            context.SaveChanges();
        }
    }
}
